docker/pusher.go

```diff
@@ -2,67 +2,142 @@ package docker
+}
+
+func WithCredentialsProvider(cp CredentialsProvider) Opt {
+	return func(p *Pusher) error {
+		p.credentialsProvider = cp
+		return nil
+	}
+}
+
+func EmptyCredentialsProvider(ctx context.Context, registry string) (Credentials, error) {
+	return Credentials{}, nil
 }
 
 // NewPusher creates an instance of a docker-based image pusher.
-func NewPusher() *Pusher {
-	return &Pusher{}
+func NewPusher(opts ...Opt) (*Pusher, error) {
+	result := &Pusher{
+		Verbose: false,
+		credentialsProvider: EmptyCredentialsProvider,
+	}
+	for _, opt := range opts {
+		err := opt(result)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
 }
 
 // Push the image of the Function.
-func (n *Pusher) Push(f bosonFunc.Function) (digest string, err error) {
-	// Check for the docker binary explicitly so that we can return
-	// an extra-friendly error message.
-	_, err = exec.LookPath("docker")
-	if err != nil {
-		err = errors.New("please install 'docker'")
-		return
-	}
+func (n *Pusher) Push(ctx context.Context, f bosonFunc.Function) (digest string, err error) {
 
 	if f.Image == "" {
 		return "", errors.New("Function has no associated image.  Has it been built?")
 	}
 
-	// set up the command, specifying a sanitized project name and connecting
-	// standard output and error.
-	cmd := exec.Command("docker", "push", f.Image)
+	var registry string
+	parts := strings.Split(f.Image, "/")
+	switch len(parts) {
+	case 2:
+		registry = "docker.io"
+	case 3:
+		registry = parts[0]
+	default:
+		return "", errors.Errorf("failed to parse image name: %q", f.Image)
+	}
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create docker api client")
+	}
+
+	credentials, err := n.credentialsProvider(ctx, registry)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get credentials")
+	}
+
+	b, err := json.Marshal(&credentials)
+	if err != nil {
+		return "", err
+	}
+
+	opts := types.ImagePushOptions{RegistryAuth: base64.StdEncoding.EncodeToString(b)}
 
-	// Capture the command output in the buffer
-	var output bytes.Buffer
+	r, err := cli.ImagePush(ctx, f.Image, opts)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to push the image")
+	}
+	defer r.Close()
+
+	var output io.Writer
+	var outBuff bytes.Buffer
 
 	// If verbose logging is enabled, echo chatty stdout.
 	if n.Verbose {
-		fmt.Println(cmd)
-		cmd.Stdout = io.MultiWriter(&output, os.Stdout)
+		output = io.MultiWriter(&outBuff, os.Stdout)
 	} else {
-		cmd.Stdout = &output
+		output = &outBuff
 	}
 
-	// Capture stderr for echoing on failure.
-	var stderr bytes.Buffer
-	cmd.Stderr = &stderr
-
-	// Run the command, echoing captured stderr as well ass the cmd internal error.
-	err = cmd.Run()
-	if err != nil {
-		// TODO: sanitize stderr from docker?
-		err = fmt.Errorf("%v. %v", stderr.String(), err.Error())
+	decoder := json.NewDecoder(r)
+	li := logItem{}
```

Comment:
Output of this is not so nice as of `docker` cli, but I think that's OK. I would need to deal with cursor moving I don't want to spend time on that.